### PR TITLE
ENH: Initialize global pause render state for newly created views

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
Previously, newly created views were initialized with the default pause render state of 0. This caused warning messages if the layout was changed between pauseRender/resumeRender calls as the render state of newly created views would not match the existing ones.

This commit tracks the global pauseRender count, and uses it to initialize the pauseRender state of newly created views. This removes most instances of the warning message:
> [Qt] int __cdecl ctkVTKAbstractView::resumeRender(void) Cannot resume rendering, pause render count is already 0!

### CODE REVIEW

File: Libs/MRML/Widgets/qMRMLLayoutManager.cxx [hunk 1]
```diff
@@ -260,6 +263,10 @@ QWidget* qMRMLLayoutSliceViewFactory::createViewFromNode(vtkMRMLAbstractViewNode
   sliceWidget->setMRMLScene(this->mrmlScene());
   sliceWidget->setSliceLogics(this->sliceLogics());
   this->sliceLogics()->AddItem(sliceWidget->sliceLogic());
+  for (int i = 0; i < this->layoutManager()->globalPauseRenderCount(); ++i)
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLLayoutManager.cxx:266]: This should be done for all new views (not just for 3D and slice views). Could we do it at one place, right after we creating the view, in qMRMLLayoutViewFactory::onViewNodeAdded?

File: Libs/MRML/Widgets/qMRMLLayoutManager.cxx [hunk 2]
```diff
@@ -1244,6 +1237,15 @@ void qMRMLLayoutManager::setRenderPaused(bool pause)
     ctkVTKAbstractView* view = this->threeDWidget(viewName)->threeDView();
     view->setRenderPaused(pause);
     }
+
+  if (pause)
+    {
+    d->GlobalPauseRenderCount += 1;
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLLayoutManager.cxx:1243]:
```suggestion
    d->GlobalPauseRenderCount++;
```

File: Libs/MRML/Widgets/qMRMLLayoutManager.cxx [hunk 3]
```diff
@@ -1244,6 +1237,15 @@ void qMRMLLayoutManager::setRenderPaused(bool pause)
...
+    }
+  else
+    {
+    d->GlobalPauseRenderCount = std::max(d->GlobalPauseRenderCount - 1, 0);
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLLayoutManager.cxx:1247]: Let's print a warning if it goes below 0.

File: Libs/MRML/Widgets/qMRMLLayoutManager_p.h [hunk 4]
```diff
@@ -155,6 +155,7 @@ public slots:
   /// Top-level windows created and managed by this class that host additional viewports
   /// (outside the main application window).
   QMap<QString, ViewportInfo>  DetachedViewports;
+  int GlobalPauseRenderCount{ 0 };
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLLayoutManager_p.h:158]: It is not quite global, just all the views that this layout manager owns. Could you rename it to "AllViewsPauseRenderCount" or something similar?

File: Libs/MRML/Widgets/qMRMLLayoutManager.cxx [hunk 5]
```diff
@@ -1221,21 +1228,7 @@ void qMRMLLayoutManager::setRenderPaused(bool pause)
...
-      // and just log a debug message here to help with troubleshooting if
-      // any problem comes up related to pausing rendering.
-      qDebug() << Q_FUNC_INFO << "Resume render request is ignored for view "
-        << viewName << ", probably the view has just been created";
-      }
-    else
-      {
-      view->setRenderPaused(pause);
-      }
+    view->setRenderPaused(pause);
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLLayoutManager.cxx:1231]: Please still keep the warning message if trying to unpause render on a view that has no paused rendering on it. Thanks to your fixes, we should no longer see the message.

**Review by lassoan — COMMENTED** [review]
Thank you, this will be very good. I've added a few comments inline.

File: Libs/MRML/Widgets/qMRMLLayoutManager.cxx (hunk 5 above)
Comment (Sunderlandkyl) [reply Libs/MRML/Widgets/qMRMLLayoutManager.cxx:1231]: This isn't the warning message that we are used to seeing, This is a debug message that was hiding the warning message for slice views. The actual warning message is still in CTK: https://github.com/commontk/CTK/blob/master/Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp#L285.

Comment (lassoan) [reply Libs/MRML/Widgets/qMRMLLayoutManager.cxx:1231]: It is very good then, thanks for the clarification.

File: Libs/MRML/Widgets/qMRMLLayoutViewFactory.cxx [hunk 6]
```diff
@@ -407,23 +406,13 @@ void qMRMLLayoutViewFactory::onViewNodeAdded(vtkMRMLAbstractViewNode* node)
...
-    view = threeDWidget->threeDView();
-    }
-  qMRMLSliceWidget* sliceWidget = qobject_cast<qMRMLSliceWidget*>(viewWidget);
-  if (sliceWidget)
-    {
-    view = sliceWidget->sliceView();
-    }
-
-  if (view)
+  qMRMLAbstractViewWidget* abstractView = qobject_cast<qMRMLAbstractViewWidget*>(viewWidget);
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLLayoutViewFactory.cxx:409]: it is not a view but a widget, so a name like "viewWidget" would be better

File: Libs/MRML/Widgets/qMRMLPlotWidget.cxx [hunk 7]
```diff
@@ -116,37 +119,47 @@ void qMRMLPlotWidget::setMRMLPlotViewNode(vtkMRMLPlotViewNode* newPlotViewNode)
   d->PlotController->setMRMLPlotViewNode(newPlotViewNode);
 }
 
-// --------------------------------------------------------------------------
+//--------------------------------------------------------------------------
+void qMRMLPlotWidget::setMRMLAbstractViewNode(vtkMRMLAbstractViewNode* newPlotViewNode)
+{
+  this->setMRMLPlotViewNode(vtkMRMLPlotViewNode::SafeDownCast(newPlotViewNode));
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLPlotWidget.cxx:125]: we should log a warning if a wrong view node type is provided (not just here but in all other widget classes, too).

File: Libs/MRML/Widgets/qMRMLTableWidget.cxx [hunk 8]
```diff
@@ -133,22 +149,22 @@ qMRMLTableView* qMRMLTableWidget::tableView()const
...
-//---------------------------------------------------------------------------
-QString qMRMLTableWidget::viewLabel()const
+// --------------------------------------------------------------------------
+qMRMLViewControllerBar* qMRMLTableWidget::controllerWidget()const
 {
   Q_D(const qMRMLTableWidget);
-  return d->TableController->viewLabel();
-}
+  return this->tableController();
+}
```
Comment (lassoan) [inline Libs/MRML/Widgets/qMRMLTableWidget.cxx:170]: newline is missing at the end?

**Review by lassoan — COMMENTED** [review]
Thank you, this looks great, I've just added a few trivial comments inline.

File: Libs/MRML/Widgets/qMRMLLayoutViewFactory.cxx (hunk 6 above)
Comment (Sunderlandkyl) [reply Libs/MRML/Widgets/qMRMLLayoutViewFactory.cxx:409]: "viewWidget" is already in use for the QWidget* variable. How about "abstractViewWidget"?

Comment (lassoan) [reply Libs/MRML/Widgets/qMRMLLayoutViewFactory.cxx:409]: or maybe rename the `viewWidget` above to `widget` (as it is more generic)

Comment (Sunderlandkyl) [reply Libs/MRML/Widgets/qMRMLLayoutViewFactory.cxx]: Ok, I've renamed it "viewWidget"->"widget", and "abstractView"->"viewWidget".

**Review by lassoan — APPROVED** [review]
This looks good, thank you!

**Pull request merged**